File: apps/mobile/lib/services/connection_url_parser.dart
sealed class DeepLinkParams {}

class ConnectionParams extends DeepLinkParams {
  final String serverUrl;
  final String? token;

  ConnectionParams({required this.serverUrl, this.token});
}

class SessionLinkParams extends DeepLinkParams {
  final String sessionId;

  SessionLinkParams({required this.sessionId});
}

class ConnectionUrlParser {
  /// Parses a deep link URL into [DeepLinkParams].
  ///
  /// Supported formats:
  /// - `ccpocket://connect?url=ws://IP:PORT&token=...` → [ConnectionParams]
  /// - `ccpocket://session/<sessionId>` → [SessionLinkParams]
  /// - `ws://IP:PORT` or `wss://IP:PORT` → [ConnectionParams]
  /// - `IP:PORT` (treated as ws://) → [ConnectionParams]
  static DeepLinkParams? parse(String rawUrl) {
    final trimmed = rawUrl.trim();
    if (trimmed.isEmpty) return null;

    // Deep link: ccpocket://...
    if (trimmed.startsWith('ccpocket://')) {
      final uri = Uri.tryParse(trimmed);
      if (uri == null) return null;

      // ccpocket://session/<sessionId>
      if (uri.host == 'session') {
        final segments = uri.pathSegments;
        if (segments.isEmpty) return null;
        final sessionId = segments.first;
        if (sessionId.isEmpty) return null;
        return SessionLinkParams(sessionId: sessionId);
      }

      // ccpocket://connect?url=...&token=...
      if (uri.host == 'connect') {
        final url = uri.queryParameters['url'];
        if (url == null || url.isEmpty) return null;
        final token = uri.queryParameters['token'];
        return ConnectionParams(
          serverUrl: url,
          token: (token != null && token.isNotEmpty) ? token : null,
        );
      }

      return null;
    }

    // Direct ws:// or wss://
    if (trimmed.startsWith('ws://') || trimmed.startsWith('wss://')) {
      return ConnectionParams(serverUrl: trimmed);
    }

    // Bare host:port
    final hostPortPattern = RegExp(r'^[\w.\-]+:\d+$');
    if (hostPortPattern.hasMatch(trimmed)) {
      return ConnectionParams(serverUrl: 'ws://$trimmed');
    }

    return null;
  }
}

File: iOS/CCPocketNative/CCPocketNative/Services/ConnectionURLParser.swift
import Foundation

enum DeepLinkParams: Equatable {
    case connection(ConnectionParams)
    case session(String)
}

enum ConnectionURLParser {
    static func parse(_ rawURL: String) -> DeepLinkParams? {
        let trimmed = rawURL.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }

        if trimmed.hasPrefix("ccpocket://") {
            guard let components = URLComponents(string: trimmed) else { return nil }
            if components.host == "session" {
                let sessionId = components.path.split(separator: "/").first.map(String.init)
                return sessionId.map(DeepLinkParams.session)
            }

            if components.host == "connect" {
                let url = components.queryItems?.first(where: { $0.name == "url" })?.value
                let token = components.queryItems?.first(where: { $0.name == "token" })?.value
                guard let url, isValidWebSocketURL(url) else { return nil }
                return .connection(ConnectionParams(serverURL: url, token: token?.isEmpty == false ? token : nil))
            }
            return nil
        }

        if trimmed.hasPrefix("ws://") || trimmed.hasPrefix("wss://") {
            return isValidWebSocketURL(trimmed) ? .connection(ConnectionParams(serverURL: trimmed)) : nil
        }

        if isHostPort(trimmed) || isBracketedIPv6Port(trimmed) {
            return .connection(ConnectionParams(serverURL: "ws://\(trimmed)"))
        }

        return nil
    }

    static func parse(_ url: URL) -> DeepLinkParams? {
        parse(url.absoluteString)
    }

    private static func isValidWebSocketURL(_ value: String) -> Bool {
        guard let components = URLComponents(string: value),
              components.scheme == "ws" || components.scheme == "wss",
              components.host?.isEmpty == false else {
            return false
        }
        if let port = components.port {
            return (1...65535).contains(port)
        }
        return true
    }

    private static func isHostPort(_ value: String) -> Bool {
        value.range(of: #"^[\w.\-]+:\d+$"#, options: .regularExpression) != nil
    }

    private static func isBracketedIPv6Port(_ value: String) -> Bool {
        value.range(of: #"^\[[^\]]*:[^\]]*\]:\d+$"#, options: .regularExpression) != nil
    }
}

